src/emqx_protocol.erl

```diff
@@ -421,7 +423,96 @@ process_packet(?DISCONNECT_PACKET(?RC_SUCCESS, #{'Session-Expiry-Interval' := In
+    disconnect({?RC_PROTOCOL_ERROR, PState});
+process_packet(?AUTH_PACKET(?RC_CONTINUE_AUTHENTICATE, Properties), PState) ->
+    process_enhanced_auth(maps:get('Authentication-Method', Properties, undefined), maps:get('Authentication-Data', Properties, undefined), PState);
+process_packet(?AUTH_PACKET(?RC_RE_AUTHENTICATE, Properties), PState = #pstate{auth_state = connected}) ->
+    process_enhanced_auth(maps:get('Authentication-Method', Properties, undefined), maps:get('Authentication-Data', Properties, undefined), PState);
+process_packet(?AUTH_PACKET(?RC_RE_AUTHENTICATE), PState = #pstate{auth_state = reauthenticating}) ->
+    disconnect({?RC_PROTOCOL_ERROR, PState});
+process_packet(?AUTH_PACKET(?RC_RE_AUTHENTICATE), PState = #pstate{auth_state = authenticating}) ->
+    connack({?RC_PROTOCOL_ERROR, PState}).
+
+%%------------------------------------------------------------------------------
+%% process enhanced authenticate
+%%------------------------------------------------------------------------------
+
+process_enhanced_auth(AuthMethod, _AuthData, PState = #pstate{auth_state = AuthState}) 
+    when AuthMethod =:= undefined orelse AuthMethod =:= <<>> ->
+    if 
+        AuthState =:= reauthenticating orelse AuthState =:= connected ->
+            disconnect({?RC_MALFORMED_PACKET, PState});
+        true ->
+            connack({?RC_MALFORMED_PACKET, PState})
+    end;
+process_enhanced_auth(AuthMethod, AuthData, PState = #pstate{auth_method = OldAuthMethod, auth_state = AuthState}) ->
+    if 
+        OldAuthMethod =/= <<>> andalso OldAuthMethod =/= AuthMethod ->
+            if 
+                AuthState =:= reauthenticating orelse AuthState =:= connected ->
+                    disconnect({?RC_BAD_AUTHENTICATION_METHOD, PState});
+                true ->
+                    connack({?RC_BAD_AUTHENTICATION_METHOD, PState})
+            end;
+        true ->
+            case authenticate(credentials(PState, #{auth_method => AuthMethod, 
+                                                    auth_data   => AuthData})) of
+                {continue, Reply} -> 
+                    deliver({auth, ?RC_CONTINUE_AUTHENTICATE, AuthMethod, Reply}, PState#pstate{auth_method = AuthMethod,
+                                                                                                auth_state = case AuthState of
+                                                                                                                 connected ->
+                                                                                                                     reauthenticating;
+                                                                                                                 _ ->
+                                                                                                                     authenticating
+                                                                                                             end});
```

Comment:
`case` and `end` should be aligned. And I suggest you to wrap your code. The code is toow long in a line to read.